packages/@react-stately/table/src/useTreeGridState.ts

```diff
@@ -0,0 +1,310 @@
+  } = opts;
+
+  let body: GridNode<T>;
+  let columns: GridNode<T>[] = [];
+  let flattenedRows = [];
+  let userColumnCount = 0;
+  let originalColumns = [];
+  let keyMap = new Map();
+
+  if (opts?.showSelectionCheckboxes) {
+    let rowHeaderColumn: GridNode<T> = {
+      type: 'column',
+      key: ROW_HEADER_COLUMN_KEY,
+      value: null,
+      textValue: '',
+      level: 0,
+      index: opts?.showDragButtons ? 1 : 0,
+      hasChildNodes: false,
+      rendered: null,
+      childNodes: [],
+      props: {
+        isSelectionCell: true
+      }
+    };
+
+    columns.unshift(rowHeaderColumn);
+  }
+
+  // Add cell for drag buttons if needed.
+  if (opts?.showDragButtons) {
+    let rowHeaderColumn: GridNode<T> = {
+      type: 'column',
+      key: ROW_HEADER_COLUMN_KEY_DRAG,
+      value: null,
+      textValue: '',
+      level: 0,
+      index: 0,
+      hasChildNodes: false,
+      rendered: null,
+      childNodes: [],
+      props: {
+        isDragButtonCell: true
+      }
+    };
+
+    columns.unshift(rowHeaderColumn);
+  }
+
+  let topLevelRows = [];
+  let visit = (node: GridNode<T>) => {
+    switch (node.type) {
+      case 'body':
+        body = node;
+        keyMap.set(body.key, body);
+        break;
+      case 'column':
+        if (!node.hasChildNodes) {
+          columns.push(node);
+          userColumnCount++;
+        }
+        break;
+      case 'item':
+        topLevelRows.push(node);
+        return;
+    }
+
+    for (let child of node.childNodes) {
+      visit(child);
+    }
+  };
+
+  for (let node of nodes) {
+    if (node.type === 'column') {
+      originalColumns.push(node);
+    }
+    visit(node);
+  }
+
+  // Update each grid node in the treegrid table with values specific to a treegrid structure. Also store a set of flattened row nodes for TableCollection to consume
+  let globalRowCount = 0;
+  let visitNode = (node: GridNode<T>, i?: number) => {
+    // Clone row node and its children so modifications to the node for treegrid specific values aren't applied on the nodes provided
+    // to TableCollection. Index, level, and parent keys are all changed to reflect a flattened row structure rather than the treegrid structure
+    // values automatically calculated via CollectionBuilder
+    if (node.type === 'item') {
+      let childNodes = [];
+      for (let child of node.childNodes) {
+        if (child.type === 'cell') {
+          let cellClone = {...child};
+          if (cellClone.index + 1 === columns.length) {
+            cellClone.nextKey = null;
+          }
+          childNodes.push({...cellClone});
+        }
+      }
+      let clone = {...node, childNodes: childNodes, parentKey: body.key, level: 1, index: globalRowCount++};
+      flattenedRows.push(clone);
+    }
+
+    let newProps = {};
+    // Associate the parent column node to the cell/column cell.
+    if (node.type === 'cell' || node.type === 'column') {
+      newProps['column'] = columns[i];
```

Comment:
I wonder if we can avoid assigning column to the treegrid nodes and instead rely on the flattened nodes to have this property? It should already be assigned by TableCollection there. That would allow you to remove the duplicate logic above that finds the columns, adds the checkbox column, etc. Should be easy enough to find the corresponding flattened node by key whenever you need the `column`?

Reply:
Hmm, good point, at a glance I think I can do that. I think I was more concerned with just having the tree grid key map data having as much parity as possible to the TableCollection nodes + having fully fleshed out data for the cell nodes in the tree grid key map (aka calculating indexOfType/nextKey/etc) which all required adding to the `columns` array here with the checkbox column

Reply:
Pushed an update removing the column node assignment to the treegrid nodes. Preserved the `columns` array just so it is easier to determine when a treegrid cell is the last cell in the row and add `nextKey: null`.

Reply:
Oh you probably didn't need an array for that, you could have a `columnCount` that you add the number of extra columns to?